### DISCUSSION TITLE
Question on fine-grained synchronization in MiCS

### DISCUSSION THREAD

**Discussion by learning-chip** [deepspeedai/DeepSpeed#3896] [Q&A] [answered]
In the [MiCS paper](https://dl.acm.org/doi/10.14778/3561261.3561265), Section 4, "Fine-grained synchronization", it is stated that 

> MiCS follows the good practice in existing works, e.g., BytePS [26], that leverages much finer-grained wait_event, wait_stream and record_stream operations for synchronization, which allow us to maintain the relative order of computation and communication operations in different streams.

However, in the implementation #2964, I cannot find invocation of `wait_event`, `wait_stream` or `record_stream`. Is it not implemented yet? How large is the performance impact? Thanks!

**Comment by tjruwase** [comment]
@zarzen, FYI

**Comment by zarzen** [comment, marked as answer]
Thanks @tjruwase for mentioning. 

@learning-chip Thanks for being interested in our work, the latest DeepSpeed implementation for parameter gathering has already included these fine-grained event control. You can view it as implemented already.

**Reply by learning-chip** [reply]
Thanks! Just to confirm: do you mean this `reduce_and_partition_stream.wait_stream` inside `DeepSpeedZeroOptimizer_Stage3`:

https://github.com/microsoft/DeepSpeed/blob/7e8bcc07d6532a380a7a439741609b058f083ed6/deepspeed/runtime/zero/stage3.py#L1085

which is then inherited by [class MiCS_Optimizer](https://github.com/microsoft/DeepSpeed/blob/7e8bcc07d6532a380a7a439741609b058f083ed6/deepspeed/runtime/zero/mics.py#L351)?

**Reply by zarzen** [reply]
Yes it is
